conctr.agent.lib.nut

```diff
@@ -195,18 +228,393 @@ class Conctr {
+        headers["Authorization"] <- _headers["Authorization"];
+        payload["topics"] <- topics;
+        payload["clientId"] <- _device_id;
+        
+        local _oldReq = null;
+        
+        // Check there isnt a current connection, if there is keep a copy
+        if (_pollingReq) _oldReq = _pollingReq;
+        if(_wakeAndSubTimer) imp.cancelwakeup(_wakeAndSubTimer);
+        
+        _wakeAndSubTimer = imp.wakeup(CONN_TIMEOUT, reconnect.bindenv(this))
+
+        _pollingReq = http.post(_pubSubEndpoints[action], headers, http.jsonencode(payload));
+
+        // Call callback directly when not chucked response, handle chuncking in second arg to sendAsync
+        _pollingReq.sendasync(_doneCb.bindenv(this), _streamCb.bindenv(this), CONN_TIMEOUT+10);
+        
+        // new req made, cancel old req
+        if(_oldReq) _oldReq.cancel();
+    }
+
+
+    // 
+    // Unsubscribe to a single/list of topics
+    // 
+    function unsubscribe() {
+        if (_pollingReq) _pollingReq.cancel();
+        if(_wakeAndSubTimer) imp.cancelwakeup(_wakeAndSubTimer);
+        _pollingReq = null;
+    }
+
+
+    // 
+    // Http GET request with conctr auth injected in automatically
+    // 
+    // @param  {String}     url       Url to hit
+    // @param  {Table}      headers   Additional headers pass
+    // @param  {Function}   cb        cb called with result
+    // 
+    function get(url, headers = {}, cb = null) {
+        if (typeof headers == "function") {
+            cb = headers;
+            headers = {};
+        }
+
+        headers["Authorization"] <- _headers["Authorization"];
+
+        _requestWithRetry("GET", url, headers, "", function(err, resp) {
+            if (err) cb(err,null);
+            else cb(null, resp);
+        }.bindenv(this));
+    }
+
+
+    // 
+    // Http POST request with conctr auth injected in automatically
+    // 
+    // @param  {String}     url       Url to hit
+    // @param  {Table}      payload   Payload to send
+    // @param  {Table}      headers   Additional headers pass
+    // @param  {Function}   cb        cb called with result
+    // 
+    function post(url, payload, headers = {}, cb = null) {
+        if (typeof headers == "function") {
+            cb = headers;
+            headers = {};
+        }
+
+        if (typeof payload != "string") {
+            payload = http.jsonencode(payload);
+        }
+
+        headers["Authorization"] <- _headers["Authorization"];
+
+        _requestWithRetry("POST", url, headers, payload, function(err, resp) {
+            if (err) cb(err);
+            else cb(resp);
+        }.bindenv(this));
+    }
+
+    /**
+     * Logs to conctr logs and also to server.log
+     * 
+     * @param  {String} msg Message to store. If type is not string it will be json encoded.
+     */
+    function log(msg){
+
+        if (typeof msg != "string") {
+            try {
+                msg = http.jsonencode(msg);
```

Comment:
I don't think `http.jsonencode()` throws exceptions. And if it does ... let the develop catch it.  Logging an error serves no purpose.